### PR TITLE
pgroonga-primary-maintainer: Add a parameter that can limit run hours

### PR DESCRIPTION
**Pull request by abetomo**
Allows it to run only at specific hours to avoid load.

### CODE REVIEW

File: src/pgroonga-primary-maintainer.c [hunk 1]
```diff
@@ -209,6 +213,83 @@ pgroonga_primary_maintainer_wal_size_check_all(void)
 	pgstat_report_activity(STATE_IDLE, NULL);
 }
 
+static bool
+checkPGroongaPrimaryMaintainerHours(char **newval,
```
Comment (kou) [inline src/pgroonga-primary-maintainer.c:217]:
```suggestion
PGroongaPrimaryMaintainerHoursCheck(char **newValue,
```

File: src/pgroonga-primary-maintainer.c [hunk 2]
```diff
@@ -209,6 +213,83 @@ pgroonga_primary_maintainer_wal_size_check_all(void)
...
+		return true;
+
+	rawstring = pstrdup(*newval);
+	if (!SplitDirectoriesString(rawstring, ',', &elemlist))
+	{
+		list_free_deep(elemlist);
+		pfree(rawstring);
+		ereport(LOG,
+				(errcode(ERRCODE_SYNTAX_ERROR),
+				 errmsg("invalid syntax in pgroonga_primary_maintainer.hour")));
```
Comment (kou) [inline src/pgroonga-primary-maintainer.c:242]: Is this correct?
It seems that we need to call `GUC_check_err*()` instead:
* https://github.com/postgres/postgres/blob/c61c0cb3a99285db36806b592bd8c540c98757e5/src/include/utils/guc.h#L443-L453
* https://github.com/postgres/postgres/blob/c61c0cb3a99285db36806b592bd8c540c98757e5/src/backend/replication/syncrep.c#L1009-L1014

File: src/pgroonga-primary-maintainer.c [hunk 3]
```diff
@@ -209,6 +213,83 @@ pgroonga_primary_maintainer_wal_size_check_all(void)
...
+	for (i = 0; i < HOURS_SIZE; i++)
+	{
+		hours[i] = false;
+	}
+
+	if (*newval == NULL || *newval[0] == '\0')
+		return true;
+
+	rawstring = pstrdup(*newval);
+	if (!SplitDirectoriesString(rawstring, ',', &elemlist))
```
Comment (kou) [inline src/pgroonga-primary-maintainer.c:236]: Hmm...
Can we generalize `PGrnOptionParseNames()` and use it instead?
We don't want to allocate memory for this.

File: src/pgroonga-primary-maintainer.c [hunk 4]
```diff
@@ -209,6 +213,83 @@ pgroonga_primary_maintainer_wal_size_check_all(void)
...
+		hours[hour] = true;
+	}
+
+	list_free_deep(elemlist);
+	pfree(rawstring);
+
+	// for 16 or higher
+	// *extra = guc_malloc(LOG, HOURS_SIZE * sizeof(bool));
+	*extra = malloc(HOURS_SIZE * sizeof(bool));
+	memcpy(*extra, hours, HOURS_SIZE * sizeof(bool));
```
Comment (kou) [inline src/pgroonga-primary-maintainer.c:271]: We don't want to allocate this.
Can we define `PGroongaPrimaryMaintainerHours` as `static bool PGroongaPrimaryMaintainerHours[HOURS_SIZE]` instead?
We can re-parse the user input in the assign function.

File: src/pgroonga-primary-maintainer.c [hunk 5]
```diff
@@ -25,12 +26,15 @@ extern PGDLLEXPORT void pgroonga_primary_maintainer_main(Datum datum)
 	pg_attribute_noreturn();
 
 #define TAG "pgroonga: primary-maintainer"
+#define HOURS_SIZE 24
```
Comment (kou) [inline src/pgroonga-primary-maintainer.c:29]:
```suggestion
#define PGRN_PRIMARY_MAINTAINER_HOURS_SIZE 24
```

File: src/pgroonga-primary-maintainer.c [hunk 6]
```diff
@@ -209,6 +213,83 @@ pgroonga_primary_maintainer_wal_size_check_all(void)
...
+
+	// for 16 or higher
+	// *extra = guc_malloc(LOG, HOURS_SIZE * sizeof(bool));
+	*extra = malloc(HOURS_SIZE * sizeof(bool));
+	memcpy(*extra, hours, HOURS_SIZE * sizeof(bool));
+	return true;
+}
+
+static void
+assignPGroongaPrimaryMaintainerHours(const char *newval, void *extra)
```
Comment (kou) [inline src/pgroonga-primary-maintainer.c:276]:
```suggestion
PGroongaPrimaryMaintainerHoursAssign(const char *newValue, void *extra)
```

File: src/pgroonga-primary-maintainer.c [hunk 7]
```diff
@@ -209,6 +213,83 @@ pgroonga_primary_maintainer_wal_size_check_all(void)
...
+{
+	PGroongaPrimaryMaintainerHours = (bool *) extra;
+}
+
+static bool
+ExecutionHours()
```
Comment (kou) [inline src/pgroonga-primary-maintainer.c:282]:
```suggestion
pgroonga_primary_maintainer_is_executable()
```

File: src/pgroonga-primary-maintainer.c [hunk 8]
```diff
@@ -220,6 +301,7 @@ pgroonga_primary_maintainer_main(Datum arg)
 	elog(LOG,
 		 TAG ": reindex_threshold=%d",
 		 PGroongaPrimaryMaintainerReindexThreshold);
+	elog(LOG, TAG ": hours=%s", PGroongaPrimaryMaintainerHoursString);
```
Comment (kou) [inline src/pgroonga-primary-maintainer.c:304]: Is this safe for `NULL`?

**Comment by abetomo** [comment]
Closes `pgroonga-primary-maintainer` implementation because it will stop.

**Pull request closed without merging**